File: homework_prompts/rb/w2_d3_graph.rb
#  Homework 13 - Graph
#
#  Problem: Graph
#
#  Prompt: Create a directed graph class using adjacency list edge
#          representation.
#
#            storage: {Hash Table} - the keys represent unique vertex ids and
#                     the values are arrays of Integers representing the vertex
#                     ids of its neighors.
#
#  The Graph will also have the following methods:
#
#           add_vertex: Method that accepts a vertex id {Integer} and adds the
#                      vertex to the graph.  Return true if success and false
#                      if failed.
#
#                      Input:    id {Integer}
#                     Output:    {Boolean}
#
#        remove_vertex: Method that accepts a vertex id and removes the vertex
#                      with the matching id. Return true if success and false
#                      if failed.
#
#                      Input:    id {Integer}
#                     Output:    {Boolean}
#
#             add_edge: Method that accepts two vertex ids and creates a
#                      directed edge from the first vertex to the second.
#                      Returns true if success and false if failed.
#
#                      Input:    id1 {Integer}
#                      Input:    id2 {Integer}
#                     Output:    {Boolean}
#
#          remove_Edge: Method that accepts two vertex id's and removes the
#                      directed edge from the first vertex to the second.
#                      Returns true if success and false if failed.
#
#                      Input:    id1 {Integer}
#                      Input:    id2 {Integer}
#                     Output:    {Boolean}
#
#           is_vertex:  Method that accepts an id, and returns whether the vertex
#                      exists in the graph.
#
#                      Input:    id {Integer}
#                      Output:   {Boolean}
#
#           neighbors: Method that accepts a vertex id, and returns all of the
#                      edges of that vertex.
#
#                      Input:    id {Integer}
#                      Output:   {Boolean}
#
#  Input:  {Nil}
#  Output: {Graph}
#
#  Notes:  The notation for Time/Auxiliary Space Complexity for graphs is
#          slightly different since certain functions depend on either the
#          either the number of vertices, edges, or both
#
#          O(V) = Linear w/ respect to the number of vertices
#          O(E) = Linear w/ respect to the number of all edges
#          O(V+E) = Linear w/ respect to the number of vertices plus edges
#


class Graph

  # Time Complexity:
  # Auxiliary Space Complexity:
  def add_vertex(id)
    # YOUR WORK HERE
  end

  # Time Complexity:
  # Auxiliary Space Complexity:
  def remove_vertex(id)
    # YOUR WORK HERE
  end


  # Time Complexity:
  # Auxiliary Space Complexity:
  def add_edge(id1, id2)
    # YOUR WORK HERE
  end

  # Time Complexity:
  # Auxiliary Space Complexity:
  def remove_edge(id1, id2)
    # YOUR WORK HERE
  end

  # Time Complexity:
  # Auxiliary Space Complexity:
  def is_vertex(id)
    # YOUR WORK HERE
  end

  # Time Complexity:
  # Auxiliary Space Complexity:
  def neighbors(id)
    # YOUR WORK HERE
  end
end




############################################################
###############  DO NOT TOUCH TEST BELOW!!!  ###############
############################################################

# custom expect function to handle tests
# List count : keeps track out how many tests pass and how many total
#   in the form of a two item array i.e., [0, 0]
# String name : describes the test
# Function test : performs a set of operations and returns a boolean
#   indicating if test passed
def expect(count, name, test)
  count[1] = count[1] + 1

  result = 'false'
  errMsg = nil
  begin
    if test()
      result = ' true'
      count[0] = count[0] + 1
    end
  rescue NameError => e
    MyModule::Logger.error("Error loading the deployer #{class_name}. This deployer it's not installed!")
  rescue StandardError => err
    errMsg = err.message
  end

  puts'  ' + (count[1]).to_s + ')   ' + result + ' : ' + name
  if errMsg != nil
    puts'       ' + errMsg + '\n'
  end
end


# code for checking if arrays contain the same elements
# (do not need to be in the same order)
def arrays_matching(arr1, arr2)
  if arr1.length != arr2.length
    return false
  else
    lib = {}
    for i in 0..arr1.length-1
      lib[arr1[i]] = true
    end
    for j in 0..arr2.length-1
      if lib[arr2[j]] == nil
        return false
      end
    end
    return true
  end
end


puts 'Graph Class'
test_count = [0, 0]

def test()
  work = Graph.new()
  return work.is_a?(Object)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to create an instance', test)

def test()
  work = Graph.new()
  work.instance_variable_defined?(:@storage)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has storage property', test)

def test()
  work = Graph.new()
  return work.respond_to?(:storage) && work.storage.kind_of?(Object)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'storage property initialized to an empty object', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Graph add_vertex method'
test_count = [0, 0]

def test()
  work = Graph.new()
  work.respond_to?(:add_vertex)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has add_vertex method', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  return work.respond_to?(:storage) && work.storage.length == 1
rescue StandardError => err
  puts err.message
end
expect(test_count, 'is able to add a vertex', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  return work.respond_to?(:storage) && work.storage[5].kind_of?(Array)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'verticees store an array of connections', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  return work.respond_to?(:storage) && work.storage.length == 2 && work.storage[5] && work.storage[10]
rescue StandardError => err
  puts err.message
end
expect(test_count, 'is able to add two vertices', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(5)
  return work.respond_to?(:storage) && work.storage.length == 1
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not add in duplicate vertex', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Graph add_edge method'
test_count = [0, 0]

def test()
  work = Graph.new()
  work.respond_to?(:add_edge)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has add_edge method', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  result = work.add_edge(5, 10)
  return work.respond_to?(:storage) && work.storage[5].length == 1 && work.storage[10].length == 0 && result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to create an edge between two vertices that exist', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  result = work.add_edge(5, 10)
  return work.respond_to?(:storage) && work.storage[5].length == 0 && !work.storage[10] && !result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not create an edge when one of the vertices does not exist', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Graph remove_edge method'
test_count = [0, 0]

def test()
  work = Graph.new()
  work.respond_to?(:remove_edge)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has remove_edge method', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  work.add_edge(5, 10)
  result = work.remove_edge(5, 10)
  return work.respond_to?(:storage) && work.storage[5].length == 0 && work.storage[10].length == 0 && result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to remove an edge between two vertices', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  work.add_edge(5, 10)
  result = work.remove_edge(6, 10)
  return work.respond_to?(:storage) && work.storage[5].length == 1 && !work.storage[6] && !result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not remove edge when edge does not exist', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Graph remove_vertex method'
test_count = [0, 0]

def test()
  work = Graph.new()
  work.respond_to?(:remove_vertex)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has remove_vertex method', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  result = work.remove_vertex(5)
  return work.respond_to?(:storage) && work.storage.length == 0 && result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to remove a vertex within graph', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  result = work.remove_vertex(10)
  return work.respond_to?(:storage) && work.storage.length == 1 && !result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not remove vertex that does not exist', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  work.add_vertex(15)
  work.add_edge(5, 10)
  work.add_edge(5, 15)
  work.add_edge(10, 5)
  work.add_edge(15, 10)
  result = work.remove_vertex(5)
  return work.respond_to?(:storage) && work.storage.length == 2 && work.storage[10].length == 0 && work.storage[15].length == 1 && result
rescue StandardError => err
  puts err.message
end
expect(test_count, 'removes a vertex while safely removing edges connected to node', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Graph neighbors method'
test_count = [0, 0]

def test()
  work = Graph.new()
  work.respond_to?(:neighbors)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has neighbors method', test)

def test()
  work = Graph.new()
  return work.neighbors(5) == nil
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns nil when a vertex doesn\'t exist', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  return work.neighbors(5) != nil && work.neighbors(5).length == 0
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns an empty array if vertex has no edges', test)

def test()
  work = Graph.new()
  work.add_vertex(5)
  work.add_vertex(10)
  work.add_vertex(15)
  work.add_vertex(20)
  work.add_edge(5, 10)
  work.add_edge(5, 15)
  work.add_edge(5, 20)
  result = work.neighbors(5)
  return result != nil && result.length == 3 && arrays_matching(result, [10,15,20])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns neighbors if edges exist for a vertex', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"

File: homework_prompts/swift/w2_d3_graph.swift
/*
 *  Homework 13 - Graph
 *
 *  Problem: Graph
 *
 *  Prompt: Create a directed graph class using adjacency list edge
 *          representation.
 *
 *  The Graph will have the following properties:
 *
 *             storage: {Hash Table} - the keys represent unique vertex ids
 *                      and the values are arrays of Integers representing the
 *                      vertex ids of its neighors.
 *
 *  The Graph will also have the following methods:
 *
 *           addVertex: Method that accepts a vertex id {Integer} and adds the
 *                      vertex to the graph.  Return true if success and false
 *                      if failed.
 *
 *                      Input:    id {Integer}
 *                     Output:    {Boolean}
 *
 *        removeVertex: Method that accepts a vertex id and removes the vertex
 *                      with the matching id. Return true if success and false
 *                      if failed.
 *
 *                      Input:    id {Integer}
 *                     Output:    {Boolean}
 *
 *             addEdge: Method that accepts two vertex ids and creates a
 *                      directed edge from the first vertex to the second.
 *                      Returns true if success and false if failed.
 *
 *                      Input:    id1 {Integer}
 *                      Input:    id2 {Integer}
 *                     Output:    {Boolean}
 *
 *          removeEdge: Method that accepts two vertex id's and removes the
 *                      directed edge from the first vertex to the second.
 *                      Returns true if success and false if failed.
 *
 *                       Input:    id1 {Integer}
 *                       Input:    id2 {Integer}
 *                      Output:    {Boolean}
 *
 *           isVertex:  Method that accepts an id, and returns whether the vertex
 *                      exists in the graph.
 *
 *                       Input:    id {Integer}
 *                      Output:   {Boolean}
 *
 *           neighbors: Method that accepts a vertex id, and returns all of the
 *                      edges of that vertex.
 *
 *                       Input:    id {Integer}
 *                      Output:   {Array}
 *
 *
 *  Input:  {Undefined}
 *  Output: {Graph}
 *
 *  Notes:   The notation for Time/Auxiliary Space Complexity for graphs
 *           is slightly different since certain functions depend on
 *           either the number of vertices, edges or even both
 *
 *           O(V) = Linear w/ respect to the number of vertices
 *           O(E) = Linear w/ respect to the number of edges
 *           O(V+E) = Linear w/ respect to the number of vertices plus edges
 */

class Graph {
  var storage: [Int: [Int]] = [:]

  // Time Complexity:
  // Auxiliary Space Complexity:
  func addVertex(_ id: Int) -> Bool {
  // YOUR WORK HERE
  return false
  }

  // Time Complexity:
  // Auxiliary Space Complexity:
  func addEdge(_ id1: Int,_ id2: Int) -> Bool {
  // YOUR WORK HERE
  return false
  }

  // Time Complexity:
  // Auxiliary Space Complexity:
  func removeEdge(_ id1: Int, _ id2: Int) -> Bool {
  // YOUR WORK HERE
  return false
  }

  // Time Complexity:
  // Auxiliary Space Complexity:
  func removeVertex(_ id: Int) -> Bool {
    // YOUR WORK HERE
    return false
  }

  // Time Complexity:
  // Auxiliary Space Complexity:
  func isVertex(_ id: Int) -> Bool {
  //YOUR WORK HERE
  return false
}

  // Time Complexity:
  // Auxiliary Space Complexity:
  func neighbors(_ id: Int) -> [Int]? {
  //YOUR WORK HERE
  return nil
  }
}


////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}

var testCount: [Int] = [0, 0]

print("Graph Class")

assert(&testCount, "has storage property",  {
  let g = Graph()
  return g.storage.count == 0
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]


print("Graph addVertex method")


assert(&testCount, "is able to add a single vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  return g.storage.count == 1 && g.storage[1] != nil
})


assert(&testCount, "is able to add multiple vertices",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addVertex(3)
  return g.storage.count == 3 &&
         g.storage[1] != nil &&
         g.storage[2] != nil &&
         g.storage[3] != nil
})

assert(&testCount, "does not add repeat vertices",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(1)
  return g.storage.count == 1
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]




print("Graph addEdge method")

assert(&testCount, "adds a single edge properly",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.storage.count == 2 && (g.storage[1]?.contains(2))!
})

assert(&testCount, "adds multiple edges properly",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addVertex(3)
  let _ = g.addEdge(1, 2)
  let _ = g.addEdge(1, 3)

  return g.storage.count == 3 && (g.storage[1]?.contains(2))! && (g.storage[1]?.contains(3))!
})

assert(&testCount, "does not repeat edges",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  let _ = g.addEdge(1, 2)

  return g.storage.count == 2 && (g.storage[1]?.contains(2))! && g.storage[1]?.count == 1
})

assert(&testCount, "does not add edges between 1 invalid vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 3)

  return g.storage.count == 2 && g.storage[1]?.count == 0
})

assert(&testCount, "does not add edges between 2 invalid vertices",  {
  let g = Graph()
  return g.addEdge(1, 2) == false && g.storage.count == 0
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]



print("Graph removeEdge method")


assert(&testCount, "removes a single edge properly",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.removeEdge(1, 2) == true
})

assert(&testCount, "does not remove nonexsitent edge",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  return g.removeEdge(1, 2) == false
})

assert(&testCount, "removes multiple edges",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addVertex(3)
  let _ = g.addEdge(1, 2)
  let _ = g.addEdge(1, 3)
  return g.removeEdge(1, 2) == true && g.removeEdge(1, 3) == true && g.storage[1]?.count == 0
})

assert(&testCount, "does not remove edge from 1 invalid vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.removeEdge(1, 3) == false && g.storage[1]?.count == 1
})

assert(&testCount, "does not remove edge from 2 invalid vertices",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.removeEdge(3, 4) == false && g.storage[1]?.count == 1
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]



print("Graph removeVertex method")


assert(&testCount, "removes single unconnected vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  return g.removeVertex(1) == true && g.storage.count == 0
})


assert(&testCount, "removes multiple unconnected vertices",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  return g.removeVertex(1) == true && g.removeVertex(2) == true && g.storage.count == 0
})

assert(&testCount, "removes vertex with one connection from it",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.removeVertex(1) == true && g.storage.count == 1
  // && g.neighbors(2)!.count == 0
})

assert(&testCount, "removes vertex with one connection to it",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.removeVertex(2) == true && g.storage.count == 1
  // && g.neighbors(1).count == 0
})


assert(&testCount, "removes vertex with multiple connections to it",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addVertex(3)
  let _ = g.addVertex(4)
  let _ = g.addEdge(2, 1)
  let _ = g.addEdge(3, 1)
  let _ = g.addEdge(4, 1)

  return g.removeVertex(1) == true &&
         g.storage.count == 3 &&
         g.storage[2]?.count == 0 &&
         g.storage[3]?.count == 0 &&
         g.storage[4]?.count == 0
})

assert(&testCount, "does not remove nonexsitent vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  return g.removeVertex(2) == false && g.storage.count == 1
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]



print("Graph isVertex method")

assert(&testCount, "returns true when a vertex exists",  {
  let g = Graph()
  let _ = g.addVertex(1)
  return g.isVertex(1) == true
})

assert(&testCount, "returns false when a vertex doesn't exist",  {
  let g = Graph()
return g.isVertex(1) == false
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]




print("Graph neighbors method")

assert(&testCount, "returns empty array for unconnected vertex",  {
  let g = Graph()
  let _ = g.addVertex(1)
  return g.neighbors(1)?.count == 0
})

assert(&testCount, "returns nil for nonexsitent vertex",  {
  let g = Graph()
  return g.neighbors(1) == nil
})


assert(&testCount, "works for a vertex with one edge",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addEdge(1, 2)
  return g.neighbors(1)?.count == 1 && (g.neighbors(1)?.contains(2))!
})

assert(&testCount, "works for a vertex with multiple edges",  {
  let g = Graph()
  let _ = g.addVertex(1)
  let _ = g.addVertex(2)
  let _ = g.addVertex(3)
  let _ = g.addVertex(4)
  let _ = g.addEdge(1, 2)
  let _ = g.addEdge(1, 3)
  let _ = g.addEdge(1, 4)
  return g.neighbors(1)?.count == 3
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
testCount = [0, 0]
